additional chat features/qol changes (dupe again)

yuzugamer opened the pull request:
adds alt + backspace, ctrl + backspace
adds caret/text navigation with arrow keys
makes escape close chat instead of pausing the game/exiting the lobby
reopens chat after unpausing the game

SaddestUnderscore reviewed (approved):
Lgtm, appreciate the enhancements on the chat typing

6fears7 commented on Menu/RainMeadow.MenuHooks.cs:
```diff
@@ -54,6 +55,10 @@ void IL_Menu_Update(ILContext il)
         }
+        void On_Menu_SelectNewObject(On.Menu.Menu.orig_SelectNewObject orig, Menu.Menu self, RWCustom.IntVector2 direction)
+        {
+            if(!ChatTextBox.blockInput) orig(self, direction);
```
Please wrap in an online check:

```
if (OnlineManager.lobby != null && OnlineManager.lobby.gameMode is not MeadowGameMode { 
// your code
} else {
orig(self);
}
```

6fears7 commented on Game/RainMeadow.GameHooks.cs:
```diff
@@ -106,9 +106,34 @@ private void RainWorldGame_Update(ILContext il)
+                c.Index++;
+                c.MoveAfterLabels();
+                c.Emit(OpCodes.Ldarg_0);
+                c.EmitDelegate((bool lastPause, RainWorldGame self) =>
```
Same thing for this delegate.

yuzugamer commented:
added the online checks and made the story menu still controller navigable while chat is open

SaddestUnderscore commented on Menu/RainMeadow.MenuHooks.cs:
```diff
@@ -54,6 +55,17 @@ void IL_Menu_Update(ILContext il)
+        {
+            if (OnlineManager.lobby != null && OnlineManager.lobby.gameMode is not MeadowGameMode)
+            {
+                // possibly might wanna make a remix option for disabling text navigation in story menu and enabling menu navigation?
```
has this been tested yet?

6fears7 replied:
No. Feel free to do so and give feedback

SaddestUnderscore commented:
gave this a test, enjoyed the changes this had, absolutely loved it
some feedback from people

captain redstone:
```
Lots of stuff was fixed! Love the new navigation! Love ctrl backspace! Love that the chatlog finally overcame its crippling shyness.
- Alt backspace seems like a bit of an odd keybind for clearing the chatbar, though I can't really think of a better one. Muscle memory says Ctrl A Backspace, but without text selection that couldn't work.
- I wish that escape could close the chatbar even if your input is set to controller only. Right now you can only close the chatbar with the controller's pause button; if you very suddenly need to stop typing and move, it's much easier to press escape on the way back to picking up the controller than it is to find the controller's much smaller pause button.
- Would be great if Delete and Ctrl Delete were implemented. Now that the cursor can move there's actually a point in using them.
- To a lesser degree it'd be nice if Home and End were implemented.
```

and n8rtot
```
okay small chunk of feedback; seems to work pretty consistently. Ctrl + backspace seems to delete strings fine, and alt + backspace is consistently clearing the entire message.

pressing escape to back out of chat also seems pretty consistent, havent had it pull up the actual pause screen yet despite mashing to break it.

Two features that'd be cool to see; no clue if they're planned or nah.
1. Copy/pasting: Not super important, esp seeing as chat is still quite short 
2. Shift + left/right/a to select/highlight text 


also got some weird hiccup with piggybacking but thats almost certainly unrelated. Ill go testing this build in public and see if stuff goes sideways
```

6fears7 commented:
@yuzugamer lmk if you're interested in addressing the above comments. Else we will go ahead and merge

yuzugamer commented:
oh sure. i should be able to add all of redstone's suggestions, and i might be able to add n8r's second suggestion (ctrl + shift + left/right).
i'm not inclined to add copy/pasting, because i think the chat is mostly better off without it (mainly because pasting would invite spam). though rain meadow isn't my mod, so i can totally go ahead and add it if anyone wants it to be added. should be pretty easy

SaddestUnderscore commented:
Yeah i agree with uo here, lets not have copy paste

yuzugamer commented:
added delete, home, end, text selection, made escape close chat even with a controller as the default device, and fixed pausing to leave not working in story lobby menus
